File: leetcode/216.组合总和 III.go
func combinationSum3(k int, n int) [][]int {
    ans := [][]int{}
    path := []int{}
    var dfs func(int, int)
    dfs = func(start, target int){
        if target == 0 {
            if len(path) == k {
                ans = append(ans, append([]int(nil), path...))
            }
            return
        }
        if target < 0 {
            return
        }
        for i := start; i <= 9; i++ {
            path = append(path, i)
            dfs(i + 1, target - i)
            path = path[:len(path) - 1]
        }       
    }
    dfs(1, n)
    return ans
}

File: leetcode/216.组合总和 III.py
class Solution:
    def combinationSum3(self, k: int, n: int) -> List[List[int]]:
        # # 选或者不选
        # ans = []
        # t = []
        # def dfs(i: int,target: int):
        #     if target == 0 and len(t) == k:
        #         ans.append(t[::])
        #         return
        #     if target < 0:
        #         return
        #     r = 10-i
        #     if r < k-len(t) or (i+9)*r//2 < target:
        #         return
        #     # 不选
        #     dfs(i+1, target)
        #     # 选
        #     t.append(i)
        #     dfs(i+1, target-i)
        #     t.pop()
        # dfs(1, n)
        # return ans
        ans = []
        t = []
        def dfs(i: int, target: int):
            if target < 0:
                return
            if target == 0 and len(t) == k:
                ans.append(t[::])
                return           
            r = 10-i
            if r < k-len(t) or (i+9)*r//2 < target:
                return
            for j in range(i, 10):
                t.append(j)
                dfs(j+1, target-j)
                t.pop()
        dfs(1, n)
        return ans